linkml_runtime/loaders/loader_root.py

```diff
@@ -119,7 +123,17 @@ def loads(self, source: str, target_class: Type[Union[BaseModel, YAMLRoot]], *,
         """
         return self.load(source, target_class, metadata=metadata)
 
-    def _construct_target_class(self, 
+    @abstractmethod
+    def iter_instances(self) -> Iterator[Any]:
+        """Lazily load data instances from the source
```

Comment:
[minor: can be iterated on in new PR]

Can we clarify what a data instance would be?

It seems that this is canonically a dict, *never* an instance of a class (whether dataclass or pydantic)? or would it be (e.g. pkl serialization)? Would rdflib_loader eventually implement this with a Triple/Quad object, or a 3-or-4-tuple?

I'm tending towards a more predictable signature (iterates over dicts) with some guarantees

Reply:
totally - so the type annotation here would be refined in the child loaders, keeping it "any" here is just to say "there will be some iterator (it could actually just be `Iterator` and then we would do `Iterator[dict[str, JsonObj | dict | list]]` or whatever in the child objects. We would make this type a union of all the child types but it wouldn't really give us much bc the child impl should override it